### PR TITLE
NET-1133: Update config wizard to create v1.0 configs

### PR DESCRIPTION
**Pull request by mondoreale**
## Summary

New Streamr config wizard! 🧙🏻

## Changes

The way the wizard is built is different. I still use the `inquirer` library buy flavoured slightly different. It's more about individual prompt methods (`confirm`, `select`, etc.).

I also test it different. `@inquirer/testing` helped me make the tests more "integrational". I pretend to be the user and validate the output of each scenario.

## How to run locally for a quick test?

```
npm run clean
npm run bootstrap
cd packages/broker
npx esbuild
npx esbuild bin/config-wizard.ts --platform=node --target=node18 --bundle --packages=external > dist/wizard.js && node dist/wizard.js
```

## Example output
```
✓ Congratulations, you've set up your Streamr Network node!
  Your node address is 0x394c8d686B6B083b9bd91DB9f6238140D961e922
  Your node's generated name is Deer Good Foam

  You can start your Streamr node now with
  streamr-broker /Users/mariusz/.streamr/config/default.json

  For environment specific run instructions, see
  https://docs.streamr.network/guides/how-to-run-streamr-node
```

example

### CODE REVIEW

**Comment by fonty1** [comment]
Generated config file:
Screenshot 2023-12-12 at 12 52 47
It shouldn't have this line:
❌    "$schema": "https://schema.streamr.network/config-v2.schema.json",
I.e. it should match this template:
https://docs.streamr.network/guides/become-an-operator#testnet-node-config

**Comment by mondoreale** [comment]
Looks like `npx streamr-broker-init` does it too (copied it over). Gone in [e82b4da](https://github.com/streamr-dev/network/pull/2199/commits/e82b4dab8a16e55a79764d463fbda919ae96761f).

**Review by fonty1 — APPROVED** [review]
LGTM. Probably best that @teogeb casts an eye before merging.

**Comment by mondoreale** [comment]
Current state.

https://github.com/streamr-dev/network/assets/320066/0d4d4af8-fb65-42a2-a17b-46e33eb4a8f2

Open for suggestions.

**Comment by mondoreale** [comment]
Pre-approved version (0):

example

**Comment by teogeb** [comment]
> It shouldn't have this line: ❌ "$schema": "https://schema.streamr.network/config-v2.schema.json", I.e. it should match this template: https://docs.streamr.network/guides/become-an-operator#testnet-node-config

It is good that the schema definition is there as it tells the version number of the config file and therefore allows Broker's auto-migration to be used.

Brokers work ok without the definition. For testnets the version number is not relevant as we don't plan to implement auto-migration between testnets.

**Comment by mondoreale** [comment]
@teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

**Comment by mondoreale** [comment]
Also,

```
curl https://schema.streamr.network/config-v2.schema.json
```

gives

```xml
<Error>
<script/>
<Code>AccessDenied</Code>
<Message>Access Denied</Message>
<RequestId>A60029N99F3JGEY2</RequestId>
<HostId>BupW5G6spzJcilmeTJU1tZBzsP2cynaCIlOHMq8oWXkFFLSFINn8F7zUrQp5iR7wxqAyYQ6DExY=</HostId>
</Error>
```

File: packages/broker/src/config/migration.ts [hunk 1]
```diff
@@ -144,7 +145,7 @@ export const readConfigAndMigrateIfNeeded = (fileName: string | undefined): Conf
     let explicitTargetFile = undefined
     if (fileName === undefined) {
         const defaultTargetFile = getDefaultFile()
-        const legacyTargetFile = getLegacyDefaultFile()
+        const legacyTargetFile = path.join(os.homedir(), '.streamr/broker-config.json')
```
Comment (teogeb) [inline packages/broker/src/config/migration.ts]: Why `getLegacyDefaultFile()` was inlined? Or did the functionality change?

File: packages/broker/src/config/config.ts [hunk 2]
```diff
@@ -30,20 +30,11 @@ export interface ConfigFile extends Config {
     $schema?: string
 }
 
-export interface ConfigFile extends Config {
-    $schema?: string
```
Comment (teogeb) [inline packages/broker/src/config/config.ts:34]: Revert this as `$schema` should be used (so that auto-migration works).

**Comment by mondoreale** [comment]
> @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

Oh, it doesn't fail, eh? I thought @fonty1's note came up from a schema validation failure.

Sounds like it's a good idea to bring it back. @fonty1?

File: packages/broker/bin/config-wizard.ts [hunk 3]
```diff
@@ -8,4 +8,10 @@ program
     .name('broker-config-wizard')
     .description('Run the configuration wizard for the broker')
 
-start()
+void (async () => {
```
Comment (teogeb) [inline packages/broker/bin/config-wizard.ts]:
```suggestion
;(async () => {
```

File: packages/broker/src/config/ConfigWizard.ts [hunk 4]
```diff
@@ -1,254 +1,539 @@
...
+    port?: number
 }
 
-const PLUGIN_NAMES: Record<string, string> = {
-    WS: 'websocket',
-    MQTT: 'mqtt',
-    HTTP: 'http'
+type PubsubPluginKey = 'websocket' | 'mqtt' | 'http'
+
+const DefaultPort: Record<PubsubPluginKey, number> = {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]:
```suggestion
const DEFAULT_PORTS: Record<PubsubPluginKey, number> = {
```

File: packages/broker/src/config/ConfigWizard.ts [hunk 5]
```diff
@@ -1,254 +1,539 @@
...
+    mqtt: MqttConfigSchema.properties.port.default,
+    http: BrokerConfigSchema.properties.httpServer.properties.port.default,
 }
 
-const PRIVATE_KEY_SOURCE_GENERATE = 'Generate'
-const PRIVATE_KEY_SOURCE_IMPORT = 'Import'
+/**
+ * Lets the user select and configure desired pub/sub plugins.
+ */
+async function getPubsubPlugins() {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Could add return type for all functions in this file

**Comment by fonty1** [comment]
> > @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?
> 
> Oh, it doesn't fail, eh? I thought @fonty1's note came up from a schema validation failure.
> 
> Sounds like it's a good idea to bring it back. @fonty1?

I assumed it was dropped in 1.0 since that what we've been doing to date :)

> Brokers work ok without the definition. For testnets the version number is not relevant as we don't plan to implement auto-migration between testnets.

New information to me, yeah, follow @teogeb's lead and I'll make a note of this in our Docs as well. Thanks!

File: packages/broker/src/config/ConfigWizard.ts [hunk 6]
```diff
@@ -1,254 +1,539 @@
...
+
+        if (http) {
+            Object.assign(pubsubPlugins, {
+                http: {},
+            })
+        }
+
+        const storagePath = await getStoragePath()
+
+        const httpServer = http?.port ? { port: http.port } : void 0
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]:
```suggestion
        const httpServer = http?.port ? { port: http.port } : undefined
```

File: packages/broker/src/config/ConfigWizard.ts [hunk 7]
```diff
@@ -1,254 +1,539 @@
...
+            },
+        } : {}
+
+        const { http, ...pubsubPlugins } = await getPubsubPlugins()
+
+        if (http) {
+            Object.assign(pubsubPlugins, {
+                http: {},
+            })
+        }
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Why is this needed?

File: packages/broker/src/config/ConfigWizard.ts [hunk 8]
```diff
@@ -1,254 +1,539 @@
...
-                return answers.enabledApiPlugins.includes(name)
+    writeFileSync(storagePath, JSON.stringify(config, null, 4))
+
+    chmodSync(storagePath, '600')
+}
+
+/**
+ * Adjusts the given config for the Mumbai test environment.
+ */
+function getMumbaiConfig(config: ConfigFile): ConfigFile {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: We can use the new environment preset, i.e. just defined
```
    environment: 'mumbai'
```

https://github.com/streamr-dev/network/pull/2139/files

File: packages/broker/src/config/ConfigWizard.ts [hunk 9]
```diff
@@ -1,254 +1,539 @@
...
+            plugins: {
+                ...operatorPlugins,
+                ...pubsubPlugins,
+            },
+            httpServer,
+        }
+
+        persistConfig(
+            storagePath,
+            network === 'polygon' ? config : getMumbaiConfig(config)
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Could introduce a string literal enum for networks, and also convert this ternary something like `if (network === 'mumbai') { config = getMumbaiConfig(config) }` so that possible third network will be handled correctly (if we add that at some point)

File: packages/broker/src/config/ConfigWizard.ts [hunk 10]
```diff
@@ -1,254 +1,539 @@
...
+                    log(`> ${content}`)
+                }
+            } catch (e) {
+                resume()
+
+                log('> x Failed to fetch node\'s balance')
+            }
+        }
+
+        if (operator) {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:133]: Coudl use explicit `undefined` checks here and in other if statements

**Comment by mondoreale** [comment]
I bring `$schema` back in [19d1058](https://github.com/streamr-dev/network/pull/2199/commits/19d1058fddabee3076db89d321342d89be4eb1f1).

File: packages/broker/src/config/ConfigWizard.ts [hunk 11]
```diff
@@ -1,254 +1,544 @@
...
+            > Your node address is ${chalk.greenBright(nodeAddress)}
+            > Your node's generated name is ${chalk.greenBright(getNodeMnemonic(privateKey))}
+        `)
+
+        if (operator) {
+            const resume = progress((f) =>
+                style(`> Your node address has *${f} MATIC* _– checking balance…_`)
+            )
+
+            try {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:107]: Why is this additional formating needed? Is it for removing `ETH` word?

File: packages/broker/src/config/ConfigWizard.ts [hunk 12]
```diff
@@ -1,254 +1,544 @@
...
+                log('> x Failed to fetch operator nodes')
+            }
+        }
+
+        log(`
+            >
+            > You can start your Streamr node now with
+            > *streamr-broker ${storagePath}*
+            >
+            > For environment specific run instructions, see
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:182]: Which error case this handles ? E.g. ctrl+c or something else?

File: packages/broker/src/config/ConfigWizard.ts [hunk 13]
```diff
@@ -1,254 +1,544 @@
...
+    const { data } = z
+        .object({
+            data: z.object({
+                operator: z.union([
+                    z.null(),
+                    z.object({ nodes: z.array(z.string()) }),
+                ]),
+            }),
+        })
+        .parse(await resp.json())
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:470]: Does this fix some issue with `inquirer/prompts` library?

**Review by teogeb — COMMENTED** [review]
First round of comments

**Comment by teogeb** [comment]
> @teogeb but doesn't it prevent the client from getting up? Having `$schema` in it constitutes an invalid structure from the schema validator's point of view, no?

This is a Broker config, not a client config. In Broker there is `$schema` property at top level of the config: https://github.com/streamr-dev/network/blob/c09c373a1585c7d3df01aecf7f566802fc8f1e00/packages/broker/src/config/config.schema.json#L8

**Comment by teogeb** [comment]
> schema validation failure.

Yes https://linear.app/streamr/issue/NET-1029/v2-broker-config-schema-is-not-accessible-online

File: packages/broker/bin/config-wizard.ts (hunk 3 above)
Comment (mondoreale) [reply packages/broker/bin/config-wizard.ts]: `void` is underrated. #nosemis

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: I'mma port the comments from the OG implementation…

File: packages/broker/src/config/ConfigWizard.ts (hunk 9 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: I'll reformat, but a union takes us just as far, currently. I'll keep it simple.

File: packages/broker/src/config/ConfigWizard.ts [hunk 14]
```diff
@@ -1,254 +1,544 @@
...
+        default: true,
+    })
 
-export const CONFIG_TEMPLATE: any = {
-    $schema: formSchemaUrl(CURRENT_CONFIGURATION_VERSION),
-    client: {
-        auth: {}
-    },
-    plugins: {},
-    apiAuthentication: {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:69]: Keep the `apiAuthentication` block. It should contain an auto-generated key if any API plugins (i.e. pubsub plugins) have been enabled.

**Review by teogeb — CHANGES_REQUESTED** [review]

File: packages/broker/src/config/ConfigWizard.ts (hunk 11 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:107]: `progress` prints out a busyness indicator while we're checking the wallet balance. This entire line gets erased and replaced with `Your node address has *12.34 MATIC* (…)` once we know the value.

Both texts are formatted the same way and emphasise the important part.

File: packages/broker/src/config/ConfigWizard.ts (hunk 10 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:133]: `operator` is either `undefined` or a valid ethereum address.

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (fonty1) [reply packages/broker/src/config/ConfigWizard.ts]: IIRC and IIUC :P including a plugin with default settings is done by adding an empty object like `http: {}` to the plugins section. I can't recall the exact reasoning behind it all though, perhaps we can look at documenting it together when Eric gets back?

File: packages/broker/src/config/ConfigWizard.ts (hunk 12 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:182]: Ctrl+c, yes.

File: packages/broker/src/config/ConfigWizard.ts (hunk 8 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: Very handy! Let's make proper adjustments once the env feature is live.

File: packages/broker/src/config/ConfigWizard.ts (hunk 13 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:470]: What issues?

File: packages/broker/src/config/ConfigWizard.ts (hunk 14 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:69]: Good to have you, Sir. I'll revert.

File: packages/broker/src/config/migration.ts (hunk 1 above)
Comment (mondoreale) [reply packages/broker/src/config/migration.ts]: It used to be used once, here, and assigned to `legacyTargetFile`. Does not seem to _need_ a utility function per se, no?

File: packages/broker/src/config/config.ts (hunk 2 above)
Comment (mondoreale) [reply packages/broker/src/config/config.ts:34]: Now I see why am I getting über-confused by this, haha. `ConfigFile` got de-dupped here.

Screenshot 2023-12-13 at 12 44 06 PM

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: This is the only comment that's out there. Doesn't really explain anything.
https://github.com/streamr-dev/network/blob/streamr-1.0/packages/broker/src/config/ConfigWizard.ts#L187

On Slack I described the outcome of this behaviour.

File: packages/broker/src/config/ConfigWizard.ts (hunk 10 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts:133]: Yes, therefore `if (operator !== undefined)` would communicate that check clearly

File: packages/broker/src/config/ConfigWizard.ts (hunk 8 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts]: It will be live in the same release as this config wizard

File: packages/broker/src/config/ConfigWizard.ts (hunk 11 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts:107]: Refered to `.replace(/\.(\d+)/, (f) => f.substring(0, 3))` snippet (maybe that line was the line `92` at the time of comment, or something)

File: packages/broker/src/config/ConfigWizard.ts (hunk 13 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts:470]: Ah time line numbers have changed. Maybe referred the `progress` function.

File: packages/broker/src/config/migration.ts (hunk 1 above)
Comment (teogeb) [reply packages/broker/src/config/migration.ts]: Nothing _needs_ a utility function... But we put that into a utility function so that config-related facts are in `config.ts`. If a function which gives us the default file is there, then a function that does almost the same could be also in the same place.

Maybe the actual solution would be to drop the support for legacy file location. The legacy location was used in back in Brubeck, so shouldn't be relevant in Tatum era 👍

File: packages/broker/bin/config-wizard.ts (hunk 3 above)
Comment (teogeb) [reply packages/broker/bin/config-wizard.ts]: What is the benefit of `void` here? Should we start to use this patten in `network` monorepo codebase?

The `void` is so underrated, that maybe most JS developers (me included) don't even know what this does without guessing / reading the documentation. That's one aspect why this adds unnecessary complexity.

The documentation says:
The `void` operator evaluates the given expression and then returns `undefined` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void)

Here the intention is not to evaluate any expression. Of course calling a function technically "evaluates" it as every function returns a value. But here the point is just to execute the function.

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts]: The comment is very bad, agree. The summary of Slack discussion:
- for each plugin there should be a block inside plugins block, i.e. if `httpPlugin` enabled, then there is a `http` block
- typically the non-default port is defined inside the plugin's block
  - if a default is used, no port information need to be written as we'll use defaults from the plugin schema, e.g. https://github.com/streamr-dev/network/blob/e00d704440bc68d4455c8a614442c310db9684b4/packages/broker/src/plugins/mqtt/config.schema.json#L11
- the only exception is `httpPlugin`: in that case the port is defined in the `httpServer` root level block
 - if we use default port, then no need to define the root level `httpServer` block (as empty is the default value for it in the Broker schema, and that way it uses the default port automatically) https://github.com/streamr-dev/network/blob/e00d704440bc68d4455c8a614442c310db9684b4/packages/broker/src/config/config.schema.json#L59-L60

File: packages/broker/src/config/ConfigWizard.ts (hunk 12 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts:182]: Does the throw error /force closed/ come from the inquired library? Could add a comment there

(Referring to line `172` here, line numbers corrupted again, maybe because of the force push or something)

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (teogeb) [reply packages/broker/src/config/ConfigWizard.ts]: Maybe we wouldn’t need any exceptional cases at lines `50`-`56`, i.e. we’d have just
```ts
const pubsubPlugins = await getPubsubPlugins()
```
and then when we write pubsubPlugins in plugins, we’d omit the `httpPlugin`’s port:
```ts
plugins: {
    ...operatorPlugins.map((p) => p === ‘http’ ? omit(p, ’port) : p),
    ...pubsubPlugins,
},
```
That way it would a bit clearer that this whole thing is about the port, and if getPubsubPlugins start to give other options in the future, we don’t accidentally omit all httpPlugin options at line 54 (where we currently

File: packages/broker/src/config/ConfigWizard.ts [hunk 15]
```diff
@@ -1,254 +1,591 @@
...
+    const privateKey = await (async () => {
+        if (privateKeySource === 'Generate') {
+            return Wallet.createRandom().privateKey
         }
+
+        return password({
+            message: 'Please provide the private key to import',
+            validate(value) {
+                try {
+                    return !!new Wallet(value)
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Why `!!` is used?

File: packages/broker/src/config/ConfigWizard.ts [hunk 16]
```diff
@@ -1,254 +1,591 @@
...
-    plugins: {},
-    apiAuthentication: {
-        keys: [generateApiKey()]
+        const port = Number(
+            await input({
+                message: `Provide a port for the ${key} plugin`,
+                default: defaultPort.toString(),
+                validate(value) {
+                    try {
+                        return !!z.coerce
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Why `!!` is used?

File: packages/broker/src/config/ConfigWizard.ts [hunk 17]
```diff
@@ -1,254 +1,591 @@
...
+    return privateKey
 }
 
-const generateApiKey = (): string => {
-    const hex = uuid().split('-').join('')
-    return Buffer.from(hex).toString('base64').replace(/[^0-9a-z]/gi, '')
+/**
+ * Lets the user decide the desired network for their node.
+ */
+async function getNetwork(): Promise<'polygon' | 'mumbai'> {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: Could introduce an enum for the networks

File: packages/broker/src/config/ConfigWizard.ts [hunk 18]
```diff
@@ -1,254 +1,591 @@
...
+                    } catch (e: unknown) {
+                        return (e as z.ZodError).issues
+                            .map(({ message }) => message)
+                            .join(', ')
+                    }
+                },
+            })
+        )
+
+        pubsubPlugins[key] = port !== defaultPort ? { port } : {}
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts:380]: We check if the port has already been taken at line 349. To support that, maybe it would be good to do the omitting of default ports after all the ports have been asked? 

(And it would maybe make sense to omit just the `port` from the object instead of assuming that the port is the only config option we ask. Not too relevant here, but if we add other config options, it is easier if the port omitting feature doesn't give us a surprise by omitting all options.)

File: packages/broker/src/config/ConfigWizard.ts [hunk 19]
```diff
@@ -1,254 +1,591 @@
...
+ * Gets a wallet balance of the network-native token for the given
+ * wallet address.
+ */
+async function getNativeBalance(
+    network: 'polygon' | 'mumbai',
+    address: string
+): Promise<BigNumber> {
+    const url = config[network].rpcEndpoints[0]?.url || ''
+
+    if (!/^https?:/i.test(url)) {
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]:
```suggestion
    const url = config[network].rpcEndpoints[0]?.url

    if ((url === undefined) || !/^https?:/i.test(url)) {
```

File: packages/broker/src/config/ConfigWizard.ts [hunk 20]
```diff
@@ -1,254 +1,591 @@
...
+        })
+        .parse(await resp.json())
 
-const selectStoragePath = async (): Promise<StorageAnswers> => {
-    let answers
-    do {
-        answers = await inquirer.prompt(storagePathPrompts)
-    } while (answers.overwrite === false)
-    return answers as any
+    return data.operator?.nodes || null
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]:
```suggestion
    return data.operator?.nodes ?? []
```

the return type of this function can be just `string[]` and the check at line `146` be` if (nodes.length > 0)`

File: packages/broker/src/config/ConfigWizard.ts [hunk 21]
```diff
@@ -1,254 +1,591 @@
...
+                    if (!nodes.includes(nodeAddress.toLowerCase())) {
+                        log(
+                            '> ! You will need to pair your node with your Operator:'
+                        )
+                    } else {
+                        log('> Your node has been paired with your Operator:')
+                    }
+                } else {
+                    log(`
+                        > x Your Operator could not be found on the **${network.replace(/\w/, (l) => l.toUpperCase())}** network, see
```
Comment (teogeb) [inline packages/broker/src/config/ConfigWizard.ts]: What the `network.replace(/\w/, (l) => l.toUpperCase())` does? E.g. does it capitalize the string (if that is what it does, could use https://lodash.com/docs/4.17.15#capitalize)

File: packages/broker/test/unit/ConfigWizard.test.ts [hunk 22]
```diff
@@ -1,257 +1,1689 @@
...
+    const { answers, logs }: Scenario = {
+        answers: [],
+        logs: [],
+    }
+
+    jest.spyOn(console, 'info').mockImplementation((...args: unknown[]) => {
+        const log = args
+            .join('')
+            // eslint-disable-next-line no-control-regex
+            .replace(/\x1B\[\d+m/g, '')
```
Comment (teogeb) [inline packages/broker/test/unit/ConfigWizard.test.ts]: What this does? (Removes colors, maybe?)

File: packages/broker/bin/config-wizard.ts (hunk 3 above)
Comment (mondoreale) [reply packages/broker/bin/config-wizard.ts]: Now it's an awkward argument. Let's have it your way. Doesn't make any difference here anyhway.

File: packages/broker/src/config/ConfigWizard.ts (hunk 12 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:182]: A comment here (there) is a decent idea.

File: packages/broker/src/config/ConfigWizard.ts (hunk 17 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: I'll give it a union type so I don't have to typeof it later. Enum brings no benefit.

File: packages/broker/src/config/ConfigWizard.ts (hunk 10 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:133]: It already covers the case and is sufficiently obvious. I'll leave it.

File: packages/broker/src/config/ConfigWizard.ts (hunk 7 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: That `map` is ugly af, lol. Plus, `pubsubPlugins` is an object. I got your point about the other fields we potentially lose here though. I'll get that fixed.

File: packages/broker/src/config/ConfigWizard.ts [hunk 23]
```diff
@@ -1,254 +1,591 @@
...
+            const resume = progress((f) =>
+                style(`> Your node address has *${f} MATIC* _– checking balance…_`)
+            )
+
+            try {
+                const balance = await getNativeBalance(network, nodeAddress)
+
+                const content = `Your node address has *${utils
+                    .formatEther(balance)
+                    .replace(/\.(\d+)/, (f) => f.substring(0, 3))} MATIC*`
```
Comment (mondoreale) [inline packages/broker/src/config/ConfigWizard.ts]: Oh, this? Say you have a balance of `19.7465234786528937`. This makes it `19.74`.

File: packages/broker/src/config/ConfigWizard.ts (hunk 15 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: Reformatted.

File: packages/broker/src/config/ConfigWizard.ts (hunk 16 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: Reformatted.

File: packages/broker/src/config/ConfigWizard.ts (hunk 20 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts]: What if the `operator` is null? (= meaning not found)

File: packages/broker/src/config/ConfigWizard.ts [hunk 24]
```diff
@@ -1,254 +1,607 @@
...
+            return Wallet.createRandom().privateKey
         }
+
+        return password({
+            message: 'Please provide the private key to import',
+            validate(value) {
+                try {
+                    new Wallet(value)
+
+                    return true
```
Comment (mondoreale) [inline packages/broker/src/config/ConfigWizard.ts:229]: `!!` gone.

File: packages/broker/src/config/ConfigWizard.ts [hunk 25]
```diff
@@ -1,254 +1,607 @@
...
-    plugins: {},
-    apiAuthentication: {
-        keys: [generateApiKey()]
+        const port = Number(
+            await input({
+                message: `Provide a port for the ${key} plugin`,
+                default: defaultPort.toString(),
+                validate(value) {
+                    try {
+                        z.coerce
```
Comment (mondoreale) [inline packages/broker/src/config/ConfigWizard.ts:342]: And here, `!!` gone.

File: packages/broker/src/config/ConfigWizard.ts (hunk 18 above)
Comment (mondoreale) [reply packages/broker/src/config/ConfigWizard.ts:380]: > We check if the port has already been taken at line 349. To support that, maybe it would be good to do the omitting of default ports after all the ports have been asked?

Not sure if I understand you right, but I already prevent conflicts on both custom and default ports. It's also tested.

> (And it would maybe make sense to omit just the `port` from the object instead of assuming that the port is the only config option we ask. Not too relevant here, but if we add other config options, it is easier if the port omitting feature doesn't give us a surprise by omitting all options.)

Let's be today.

File: packages/broker/src/config/migration.ts (hunk 1 above)
Comment (mondoreale) [reply packages/broker/src/config/migration.ts]: The util is back.

**Comment by mondoreale** [comment]
@teogeb the environment field is in. The PR turned a bit chaotic with all the things flying in, and the comments spread out unevenly. I don't remember gh being so… inconsistent about rebasing and comments ¯\\_(ツ)_/¯.

**Comment by mondoreale** [comment]
This is now better. I dropped the unexpected `@inquirer/testing` from the root `package.json` and kept the `@streamr/config@5.0.5` update from `streamr-1.0`.

**Review by teogeb — APPROVED** [review]
